### ISSUE TITLE
Inconsistency: input dim needed for custom modules with custom parameters but not builtin layer wrappers?

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn_common#75]
When you write some custom module (derived from `nn.Module`) which has custom parameters (created via `nn.Variable` or `nn.Parameter`), just as any PyTorch code would do ([example](https://pytorch.org/docs/stable/_modules/torch/nn/modules/activation.html#MultiheadAttention)), those are create in `__init__`. So this means that you must know the input dim at that time.

However, all the builtin layers (e.g. `Linear` etc) do not require the input dim to be known.

This is inconsistent.

How do we fix this?

Maybe the `nn.Variable` can be created more lazily and depend on the input shape? E.g. some `self.var.set_shape(...)` call in `forward`.

**Comment by albertz** [comment]
In `nn.Linear`, I have implemented a solution now which makes this optional. It creates the variable lazily. If it is not provided in `__init__`, it will be set on the first `__call__` based on the input.

I think this is ok.
